_posts/2016-08-22-the-java-synchronisers.md

```diff
@@ -0,0 +1,111 @@
+
+In this post we will explore the different options that Java provides to deal with threads synchronization.
+
+## Synchronization essentials
+
+Java provides a series of mechanisms to handle basic thread synchronization and coordination. It supports fine-grained object access synchronization through [synchronized methods](#sync-methods) and [synchronized statements](#sync-statements). Basic thread coordination can be accomplished through [guarded blocks](#guarded-blocks). All mentioned mechanisms are built around acquiring and releasing objects' intrinsic lock.
+
+### Intrisic Lock
+
+Every Java object has associated an intrinsic lock. A thread that needs exclusive access to an object's fields has to acquire the object's lock before accessing them, releasing the intrinsic lock once it is done. Other threads trying to access the object will block until the thread holding the lock releases it.
+
+### <a name="sync-methods"></a> Synchronized methods
+
+When a thread invokes a [synchronized method](https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html), it acquires the **intrinsic lock** for that method's object and releases it when the methods returns. The lock is released even if the method returns due to an uncaught exception. If done in a static method, the thread acquires the lock for the `class` object associated with the class.
+
+### <a name="sync-statements"></a> Synchronized statements
+
+Provides a more fine-grained synchronization mechanism. [Synchronized statements](https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html) must specify the object that provides the intrinsic lock. 
+
+Synchronizing over separated lock objects can provide fields synchronization, without forzing synchronization between methods calls.
+
+### <a name="guarded-blocks"></a> Guarded Blocks
+
+As mentioned earlier, [guarded blocks](https://docs.oracle.com/javase/tutorial/essential/concurrency/guardmeth.html) provide support for thread coordination. Guarded blocks are part of every Java object, through `wait`, `notify` and `notifyAll` methods.
+
+The **wait** method suspend the current thread. When a thread invokes `wait`, it must own the object's intrisic lock, that is why calls to `wait` are usually wrapped in a synchronized method or statement. The invocation of the `wait` method suspends the thread execution and releases the lock. 
+
+At some point, another thread will acquire the object's intrinsic lock and invoke `notifyAll` to inform all threads waiting that something important has happened. After the second thread has released the lock, the waiting threads will reacquire the lock and resume exectuion by returning from the `wait` invocation.
+
+**Notify** wakes up a single thread. The concrete thread can not by specified, it makes the method useful only if we do not care which threads wakes up.
```

Comment:
you might want to revist this paragraph, needs slight clarity on the explanation